lib/SILOptimizer/Utils/DeinitBarrierFinding.cpp

```diff
@@ -0,0 +1,153 @@
+//=-- DeinitBarrierFinding.cpp - Find deinit barriers above instructions. ---=//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+/// Given a list of instructions, find the deinit barriers that are above them.
+//===----------------------------------------------------------------------===//
+
+#include "swift/SILOptimizer/Utils/DeinitBarrierFinding.h"
+#include "swift/SILOptimizer/Analysis/BasicCalleeAnalysis.h"
+#include "swift/SILOptimizer/Analysis/Reachability.h"
+
+using namespace swift;
+using namespace DeinitBarrierFinding;
+
+namespace {
+
+struct Context final {
+  SILFunction &function;
+  BasicCalleeAnalysis *calleeAnalysis;
+  SILBasicBlock *defBlock;
+  ArrayRef<SILInstruction *> roots;
+};
+
+/// Works backwards from the current location of destroy_values to the earliest
+/// place they can be hoisted to.
+///
+/// Implements IterativeBackwardReachability::Effects
+/// Implements IterativeBackwardReachability::bindBarriers::Visitor
+class Dataflow final {
+  using Reachability = IterativeBackwardReachability<Dataflow>;
+  using Effect = Reachability::Effect;
+  Context const &context;
+  function_ref<bool(SILInstruction *)> isBarrier;
+  Barriers &barriers;
+  Reachability::Result result;
+  Reachability reachability;
+
+  enum class Classification { Barrier, Other };
```

Comment:
Why do we need `Classification` and `classifyInstruction` as opposed to just an `isBarrier`. It's a lot of code to understand just for one concept.

Reply:
Thanks for pointing that out!  This code was adapted from a context where it made sense to have `Classification`.  It's not worth it here.